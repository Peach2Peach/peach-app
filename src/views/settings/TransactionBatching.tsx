import { View } from "react-native";
import { Screen } from "../../components/Screen";
import { NewBubble } from "../../components/bubble/Bubble";
import { Toggle } from "../../components/inputs/Toggle";
import { useClosePopup, useSetPopup } from "../../components/popup/GlobalPopup";
import { PopupAction } from "../../components/popup/PopupAction";
import { PopupComponent } from "../../components/popup/PopupComponent";
import { ParsedPeachText } from "../../components/text/ParsedPeachText";
import { PeachText } from "../../components/text/PeachText";
import { useContractSummaries } from "../../hooks/query/useContractSummaries";
import { useSelfUser } from "../../hooks/query/useSelfUser";
import { useToggleBatching } from "../../hooks/user/useToggleBatching";
import tw from "../../styles/tailwind";
import i18n from "../../utils/i18n";
import { LoadingScreen } from "../loading/LoadingScreen";

export const TransactionBatching = () => {
  const { user, isLoading } = useSelfUser();
  const { mutate } = useToggleBatching(user || { isBatchingEnabled: false });
  const setPopup = useSetPopup();
  const closePopup = useClosePopup();
  const { contracts } = useContractSummaries();
  const toggleBatching = () => {
    const hasPendingPayouts = contracts.some(
      (contract) => contract.tradeStatus === "payoutPending",
    );
    if (user?.isBatchingEnabled && hasPendingPayouts) {
      setPopup(
        <PopupComponent
          title={i18n("settings.batching.turnOff.title")}
          content={i18n("settings.batching.turnOff.description")}
          actions={
            <>
              <PopupAction
                label={i18n("settings.batching.turnOff.no")}
                iconId="xCircle"
                onPress={closePopup}
              />
              <PopupAction
                label={i18n("settings.batching.turnOff.yes")}
                iconId="arrowRightCircle"
                onPress={() => {
                  mutate();
                  closePopup();
                }}
                reverseOrder
              />
            </>
          }
        />,
      );
    } else {
      mutate();
    }
  };
  const isBatchingEnabled = !!user?.isBatchingEnabled;

  if (isLoading) return <LoadingScreen />;
  return (
    <Screen
      style={tw`justify-center gap-8`}
      header={i18n("settings.transactionBatching")}
    >
      <View style={tw`gap-4`}>
        <PeachText style={tw`body-l`}>
          {i18n(
            isBatchingEnabled
              ? "settings.batching.delayedPayouts"
              : "settings.batching.immediatePayouts",
          )}
        </PeachText>
        <ParsedPeachText
          style={tw`body-l`}
          parse={[
            {
              pattern: new RegExp(
                i18n(
                  isBatchingEnabled
                    ? "settings.batching.youSave.highlight"
                    : "settings.batching.youPay.highlight",
                ),
                "u",
              ),
              style: tw`text-primary-main`,
            },
          ]}
        >
          {i18n.break(
            isBatchingEnabled
              ? "settings.batching.youSave"
              : "settings.batching.youPay",
          )}
        </ParsedPeachText>
        <View style={tw`flex-row items-center mt-4 gap-4`}>
          <PeachText style={tw`body-l h-6`}>
            {i18n("settings.batching.thisOnlyAppliesTo")}
          </PeachText>
          <NewBubble
            color={"black"}
            ghost={true}
            iconId="bitcoinLogo"
            disabled={true}
          >
            {i18n("wallet.wallet.bitcoin")}
          </NewBubble>
        </View>
      </View>
      <Toggle enabled={isBatchingEnabled} onPress={toggleBatching}>
        {i18n("settings.batching.toggle")}
      </Toggle>
    </Screen>
  );
};
