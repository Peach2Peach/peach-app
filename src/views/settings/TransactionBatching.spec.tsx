import { render } from "test-utils";
import { account1 } from "../../../tests/unit/data/accountData";
import { TransactionBatching } from "./TransactionBatching";

jest.mock("../../hooks/query/useSelfUser");
const useSelfUserMock = jest
  .requireMock("../../hooks/query/useSelfUser")
  .useSelfUser.mockReturnValue({
    user: account1,
    isLoading: false,
  });
describe("TransactionBatching", () => {
  it("should render correctly while loading", () => {
    useSelfUserMock.mockReturnValue({ isLoading: true });
    const { toJSON } = render(<TransactionBatching />);
    expect(toJSON()).toMatchSnapshot();
  });
  it("should render correctly", () => {
    const { toJSON } = render(<TransactionBatching />);
    expect(toJSON()).toMatchSnapshot();
  });
});
