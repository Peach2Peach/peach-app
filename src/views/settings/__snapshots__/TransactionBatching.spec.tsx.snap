// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionBatching should render correctly when batching is disabled 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -43,11 +43,11 @@
            },
            false,
          ]
        }
      >
-       Escrow payouts are delayed.
+       Escrow payouts are done right away
      </Text>
      <Text
        allowFontScaling={false}
        style={
          Array [
@@ -89,11 +89,11 @@
              ],
              undefined,
            ]
          }
        >
-         You 
+         You’re paying up to 
        </Text>
        <Text
          style={
            Array [
              Array [
@@ -116,36 +116,11 @@
                "color": "#F56522",
              },
            ]
          }
        >
-         save up to 23%
-       </Text>
-       <Text
-         style={
-           Array [
-             Array [
-               Object {
-                 "color": "#2B1911",
-                 "fontFamily": "Baloo2-Regular",
-                 "fontSize": 16,
-                 "letterSpacing": 0.16,
-                 "lineHeight": 24,
-               },
-               Object {
-                 "fontFamily": "Baloo2-Regular",
-                 "fontSize": 20,
-                 "letterSpacing": 0.2,
-                 "lineHeight": 28,
-               },
-               false,
-             ],
-             undefined,
-           ]
-         }
-       >
-          in network fees
+         30% more
        </Text>
      </Text>
    </View>
    <View
      accessibilityState={
@@ -210,11 +185,11 @@
        }
      >
        transaction batching
      </Text>
      <test-file-stub
-       fill="#F56522"
+       fill="#9F8C82"
        style={
          Array [
            Object {
              "height": 32,
              "width": 32,"
`;

exports[`TransactionBatching should render correctly when batching is enabled 1`] = `
<View
  style={
    [
      {
        "height": "100%",
        "paddingLeft": 16,
        "paddingRight": 16,
      },
      {
        "paddingLeft": 32,
        "paddingRight": 32,
      },
      {
        "gap": 32,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    style={
      {
        "gap": 16,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "color": "#2B1911",
            "fontFamily": "Baloo2-Regular",
            "fontSize": 16,
            "letterSpacing": 0.16,
            "lineHeight": 24,
          },
          {
            "fontFamily": "Baloo2-Regular",
            "fontSize": 20,
            "letterSpacing": 0.2,
            "lineHeight": 28,
          },
          false,
        ]
      }
    >
      Escrow payouts are delayed.
    </Text>
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "color": "#2B1911",
            "fontFamily": "Baloo2-Regular",
            "fontSize": 16,
            "letterSpacing": 0.16,
            "lineHeight": 24,
          },
          {
            "fontFamily": "Baloo2-Regular",
            "fontSize": 20,
            "letterSpacing": 0.2,
            "lineHeight": 28,
          },
          false,
        ]
      }
    >
      <Text
        style={
          [
            [
              {
                "color": "#2B1911",
                "fontFamily": "Baloo2-Regular",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 24,
              },
              {
                "fontFamily": "Baloo2-Regular",
                "fontSize": 20,
                "letterSpacing": 0.2,
                "lineHeight": 28,
              },
              false,
            ],
            undefined,
          ]
        }
      >
        You 
      </Text>
      <Text
        style={
          [
            [
              {
                "color": "#2B1911",
                "fontFamily": "Baloo2-Regular",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 24,
              },
              {
                "fontFamily": "Baloo2-Regular",
                "fontSize": 20,
                "letterSpacing": 0.2,
                "lineHeight": 28,
              },
              false,
            ],
            {
              "color": "#F56522",
            },
          ]
        }
      >
        save up to 23%
      </Text>
      <Text
        style={
          [
            [
              {
                "color": "#2B1911",
                "fontFamily": "Baloo2-Regular",
                "fontSize": 16,
                "letterSpacing": 0.16,
                "lineHeight": 24,
              },
              {
                "fontFamily": "Baloo2-Regular",
                "fontSize": 20,
                "letterSpacing": 0.2,
                "lineHeight": 28,
              },
              false,
            ],
            undefined,
          ]
        }
      >
         in network fees
      </Text>
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 16,
        "opacity": 1,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "color": "#2B1911",
            "fontFamily": "Baloo2-Regular",
            "fontSize": 16,
            "letterSpacing": 0.16,
            "lineHeight": 24,
          },
          [
            {
              "fontFamily": "Baloo2-SemiBold",
              "fontSize": 20,
              "letterSpacing": 0.2,
              "lineHeight": 32,
              "textTransform": "lowercase",
            },
            undefined,
          ],
          false,
        ]
      }
    >
      transaction batching
    </Text>
    <test-file-stub
      fill="#F56522"
      style={
        [
          {
            "height": 32,
            "width": 32,
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;

exports[`TransactionBatching should render correctly when loading 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <LottieAnimationView
    autoPlay={true}
    cacheComposition={true}
    colorFilters={
      [
        {
          "color": 4294272290,
          "keypath": "main",
        },
      ]
    }
    enableMergePathsAndroidForKitKatAndAbove={false}
    loop={true}
    onAnimationFinish={[Function]}
    progress={0}
    resizeMode="contain"
    sourceJson="{"v":"5.1.15","fr":60,"ip":0,"op":141,"w":120,"h":120,"nm":"Composição 1","ddd":0,"assets":[],"layers":[{"ddd":0,"ind":0,"ty":4,"nm":"main","sr":1,"ks":{"o":{"a":0,"k":100,"ix":11},"r":{"a":1,"k":[{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"n":["0p833_0p833_0p167_0p167"],"t":0,"s":[0],"e":[1080]},{"t":180}],"ix":10},"p":{"a":0,"k":[60,60,0],"ix":2},"a":{"a":0,"k":[-14.604,-15.104,0],"ix":1},"s":{"a":0,"k":[100,100,100],"ix":6}},"ao":0,"shapes":[{"ty":"gr","it":[{"d":1,"ty":"el","s":{"a":0,"k":[106.793,106.793],"ix":2},"p":{"a":0,"k":[0,0],"ix":3},"nm":"Caminho da elipse 1","mn":"ADBE Vector Shape - Ellipse","hd":false},{"ty":"st","c":{"a":0,"k":[1,1,1,1],"ix":3},"o":{"a":0,"k":100,"ix":4},"w":{"a":0,"k":11,"ix":5},"lc":2,"lj":1,"ml":4,"nm":"Traçado 1","mn":"ADBE Vector Graphic - Stroke","hd":false},{"ty":"tr","p":{"a":0,"k":[-14.604,-15.104],"ix":2},"a":{"a":0,"k":[0,0],"ix":1},"s":{"a":0,"k":[100,100],"ix":3},"r":{"a":0,"k":0,"ix":6},"o":{"a":0,"k":100,"ix":7},"sk":{"a":0,"k":0,"ix":4},"sa":{"a":0,"k":0,"ix":5},"nm":"Transformar"}],"nm":"Elipse 1","np":3,"cix":2,"ix":1,"mn":"ADBE Vector Group","hd":false},{"ty":"tm","s":{"a":1,"k":[{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"n":["0p833_0p833_0p167_0p167"],"t":0,"s":[0],"e":[0]},{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"n":["0p833_0p833_0p167_0p167"],"t":60,"s":[0],"e":[67]},{"t":119}],"ix":1},"e":{"a":1,"k":[{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"n":["0p833_0p833_0p167_0p167"],"t":0,"s":[3],"e":[70]},{"i":{"x":[0.833],"y":[0.833]},"o":{"x":[0.167],"y":[0.167]},"n":["0p833_0p833_0p167_0p167"],"t":60,"s":[70],"e":[70]},{"t":119}],"ix":2},"o":{"a":0,"k":0,"ix":3},"m":1,"ix":2,"nm":"Aparar caminhos 1","mn":"ADBE Vector Filter - Trim","hd":false}],"ip":0,"op":900,"st":0,"bm":0}],"chars":[],"markers":[]}"
    speed={1}
    style={
      [
        {
          "height": 80,
          "width": 80,
        },
        undefined,
      ]
    }
    textFiltersAndroid={[]}
    textFiltersIOS={[]}
    useNativeLooping={false}
  />
</View>
`;
